lib/llvmIr.js

```diff
@@ -0,0 +1,127 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+const utils = require('./utils');
+
+class LlvmIrParser {
+    constructor(compilerProps) {
+        // @todo More settings/compiler properties
+
+        this.maxIrLines = 500;
+        if (compilerProps) {
+            this.maxIrLines = compilerProps('maxLinesOfAsm', this.maxIrLines);
+        }
+    }
+
+    trimIrLine(line, filters) {
+        if (!filters.trim) return line;
+        const splat = line.split(/\s+/);
+        if (splat[0] === "") {
+            // An indented line: preserve a two-space indent
+            return "  " + splat.slice(1).join(" ");
+        } else {
+            return splat.join(" ");
+        }
+    }
+
+    processIr(ir, filters) {
+        const result = [];
+        let irLines = utils.splitLines(ir);
+
+        let debugInfo = {};
+
+        // @todo Get source filename ("source_filename = ...").
+
+        irLines.forEach(line => {
+            let source = null;
+            let match;
+
+            if (line.trim() === "") {
+                result.push({text: "", source: null});
+                return;
+            }
+
+            // Non-Meta IR line. Metadata is attached to it using "!dbg !123"
+            match = line.match(/!dbg (!\d+)/);
+            if (match) {
+                result.push({text: this.trimIrLine(line, filters), source: source, scope: match[1]});
+                return;
+            }
+
+            // Metadata Nodes
+            // See: https://llvm.org/docs/LangRef.html#metadata
+            // @todo: Multiline Metadata?
+            // @todo: Named Metadata?
+            match = line.match(/^(!\d+) = (distinct )?!DI([A-Za-z]+)\(([^)]+?)\)/);
+            if (match) {
+                const metaId = match[1];
+                const metaType = match[2];
+                if (!metaType) { return; }
+
+                let options = {};
+                const optionWithValue = match[4].split(',');
+                for (let i = 0; i < optionWithValue.length; ++i) {
+                    const pair = optionWithValue[i].split(':');
+                    options[pair[0].trim()] = pair[1].trim();
+                }
+                options.metaId = metaId;
+
+                if (options.line) {
+                    debugInfo[metaId] = options;
+                }
+            }
+
+            result.push({text: this.trimIrLine(line, filters), source: source});
+        });
+
+        function getLineByMetaId(metaId) {
+            const dbgInfo = debugInfo[metaId];
+            return dbgInfo ? dbgInfo.line : null;
+        }
+       
+        result.forEach(line => {
+            if (!line.scope) return;
+            // @todo: Nested Scopes?
+            line.source = {
+                file: null,
+                line: getLineByMetaId(line.scope)
+            };
+        });
+
+        return result;
+    }
+
+    process(ir, filters) {
+        return this.processIr(ir, filters);
+    }
+
+    isLlvmIr(code) {
```

Comment:
Should this be overrideable by compilers? Or do you have in mind something else to properly check for llvmir?

Reply:
I put it there so that it's available to the BaseCompiler class (in `processAsm()`). That way everything llvm related is in one place.